### ISSUE TITLE
Timeout handling under heavy load seems different between FLTK 1.4 and FLTK 1.3

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#697]
I have an application which is driven by a fixed rate timer, that is drawing a lot. Sometimes it may happen, that the time spent in the draw() routine exceeds the time slot.
I have noticed, that the same application compiled with FLTK 1.3.8 behaves much "nicer" in this situation, than with current FLTK 1.4. Nicer means, it seems to update as if the timer interval is not exceeded and from a user view things on screen stay at the intended speed. With FLTK 1.4 this is not the case, updates slow down and things on screen start to lag (from a user view).

I know that timeout handling has been completely rewritten in FLTK 1.4, so I am not sure, what the expected behaviour should be with such scenarios. But I like the FLTK 1.3 way much better and would like to have a way to have it in 1.4 too.

**To Reproduce**
I attach a short example program that tries to simulate the scenario of my application, by inserting a `sleep()` into the draw() routine:
[timer_stress_test.cxx.txt](https://github.com/fltk/fltk/files/10947748/timer_stress_test.cxx.txt)

Under FLTK 1.4 this program takes double the time to reach the same output than with FLTK 1.3.8.
I measured the time it takes to count up to 30. It takes 30 seconds with FLTK 1.3.8, but 60 seconds with FLTK 1.4.

**Expected behavior**
I would like the program to behave the same under FLTK 1.4 as with FLTK 1.3 (maybe optionally).

**FLTK Version**
Latest FLTK 1.4 from Git

**FLTK Configure / Build Options**
(Don't think that matters much, but anyway:)
`./configure --disable-wayland --disable-pango --enable-debug --with-optim="-O3"`

**Operating System / Platform:**
 - OS: Linux
 - OS Version: Ubuntu 20.04/22.04
 
**Linux/Unix Runtime, if applicable:**
 - X11

**Comment by Albrecht-S** [comment]
@wcout Thanks for the report. I can confirm the different behavior of 1.3.x (branch-1.3) and 1.4.0 (master) with the demo program. After some investigation and tests it's clear what happens, but it's not clear what can be done in FLTK's timeout and event handling. I don't think that this should be changed (see below for reasons).

The current behavior is that the timeout queue is processed and after all due timeouts have been handled this is followed by event processing before timeouts will be processed again. **Important:** Timeouts scheduled from other timeouts with `Fl::add_timeout()` or `Fl::repeat_timeout()` are **not considered** before the next event handling loop, even if the next schedule is already due.

The given demo program schedules the next timeout by calling `Fl::repeat_timeout()` when the next interval (0.01 seconds) has already expired due to longer processing of the `draw()` method. The difference between 1.3 and 1.4 is that 1.3 would process the next timeout **immediately**, i.e. before the next event handling loop. This could cause timer loops before any event handling is done and eventually lock the program in timeout processing. This had also been possible in an early version of 1.4 and caused issue #450 which was finally fixed by postponing handling of **new timers** (scheduled during timeout processing with `Fl::repeat_timeout()`) until after the next event processing loop.

The effect of the old 1.3.x behavior is that several `draw()` calls are skipped [1] and thus the program **_appears to work faster_**. However, this is not reliable (depends on system timing) and can lead to unwanted effects. This behavior was previously platform dependent and could cause very different behavior which is not what FLTK wants and why I rewrote the timeout processing to make it platform independent.

So far the current status.

**_What can be done?_**

First, [here's a patch](https://github.com/fltk/fltk/files/10948750/timeout.patch.txt) that (at least partially) restores the old behavior similar to 1.3.x. You can try it and see what it does with your original program. I guess that it could (almost) restore the old behavior but for reasons mentioned above I do not recommend to use it in production code.

[1] To show the effect I modified the given demo program and added some features for my tests. The most evident change is that it outputs the number of `draw()` calls together with the loop count while running and after termination. This demo terminates after a given number of iterations, either by default or as the first command argument: 
[timer_stress.cxx.txt](https://github.com/fltk/fltk/files/10948781/timer_stress.cxx.txt).
```
$ time bin/test/timer_stress 1000 # with patched FLTK code
FL_API_VERSION = 10400, count = 1000, draws = 500, delay = 20 ms

real	0m10,550s
```
The same program executed with FLTK 1.3:
```
$ time ./timer_stress 1000 # FLTK 1.3.x
FL_API_VERSION = 10308, count = 1000, draws = 482, delay = 20 ms

real	0m10,044s
```
As you can see the number of `draw()` calls and the runtime are similar.

I'll post a different and IMHO better solution in a follow-up comment.

**Comment by Albrecht-S** [comment]
The following solution is done in user code. This avoids the potentially faulty behavior of the timer loop and reduces the number of `draw()` calls. My recommended solution is to separate processing of calculations and data processing from display processing (i.e. drawing). I'm aware that this may need a (partial) redesign of the application structure but this will have several advantages: program stability, reliability, and predictable drawing rates ... and finally compatibility with 1.3.x and 1.4.x.

Please revert any patches before you start testing my modified test program: [timer_stress2.cxx.txt](https://github.com/fltk/fltk/files/10948818/timer_stress2.cxx.txt).

I added another constant `fps` that sets the preferred display refresh rate and calls `redraw()` in its own timer which is repeated accordingly. The widget is updated in the timer callback `cb_update()` whereas the data processing (here: counting) is done in `cb_to()` as before. The output is similar on both 1.3.x and 1.4.x and the drawing rate is very close to the requested `fps` value (not output by the program, but you can calculate it from the given data):
```
$ time bin/test/timer_stress2 1000 # FLTK 1.4.0 with original code
FL_API_VERSION = 10400, count = 1000, draws = 251, draw delay = 20 ms, fps = 20

real	0m12,561s

$ time ./timer_stress2 1000 # FLTK 1.3.x
FL_API_VERSION = 10308, count = 1000, draws = 200, draw delay = 20 ms, fps = 20

real	0m10,026s
```

Although the runtime is different, the drawing rate is very similar: about 20 fps as requested.

**_Final notes for now:_**

I'm sure that you know that drawing "as much as you can" with a 0.01 second timeout (i.e. up to 100 fps) would be wasting resources. In cases where the processing is faster than can be displayed it's probably better to limit the display rate by decoupling data processing from displaying to a rate that's consumable by a user, maybe 15 to 25 fps. I know that this can be difficult in a bigger program, but that's what I recommend anyway.

After all, the felt "behavior change" is IMHO necessary and can be different on different FLTK platforms, but I consider this as bug fixes rather than a regression. You wrote above:

> I am not sure, what the expected behaviour should be with such scenarios.

There is a conflict between event handling and timeout processing. The current algorithm makes sure that there's a kind of "snapshot" of the current timer queue that is processed before event handling continues, but newly created timer entries are considered only after the next event processing loop. Therefore, event processing is kinda "preferred" over timeout processing, and in fact you can for instance wiggle the mouse over your window and thus stall timeout processing by continuously pushing events to the event queue. That would be hard to change. I have some ideas but I don't want to pursue this until it's really necessary.

I'm open for and would appreciate comments and suggestions. Thanks for reading so far.

**Comment by MatthiasWM** [comment]
I like to use system time or another real time counter to handle skipped frames in animation correctly. This is great to keep image sequences n sync with audio and have animations run with perfect timing, even if the machine is under heavy load.

The trick is to use the Fl timeout functions, but check with `Fl::now()` and `Fl::seconds_since()` if the timer had to skip calls because frames could not be rendered in time.

https://fltk.gitlab.io/fltk/classFl.html#a9d455c09fbf3d8f34210f6d57d2977c7
https://fltk.gitlab.io/fltk/classFl.html#a908cf800c865743beda678aba01d80e2

```
int fps = 30;
Fl_Timestamp start_time;

void start_animation() {
  start_time = Fl::now();
  Fl::add_timeout(1.0/fps, render_frame);
}

void render_frame() {
  int current_frame = Fl::seconds_since(start_time) * fps;
  render_frame( current_frame );
  Fl::repeat_timeout(1.0/fps, render_frame);
}
```

**Comment by Albrecht-S** [comment]
@MatthiasWM and @wcout: In such a case - and if I interpret the given example code correctly - it would probably be better to use `Fl::add_timeout(1.0/fps, render_frame)` rather than `Fl::repeat_timeout(1.0/fps, render_frame)` in `render_frame()`. This ensures that the next frame will be scheduled exactly (as exact as possible) `1.0/fps` seconds from now and not at a time corrected by the delay of the last timeout - which is the problem of the OP. This will work fine if some timeout callbacks can execute too long and skipping frames is the better solution (as this example suggests).

**Comment by wcout** [comment]
> I like to use system time or another real time counter to handle skipped frames in animation correctly. 

@MatthiasWM Thanks for the pointer and the example! I agree, this would be the best approach.

I rewrote my test program to use it and it behaves now as it should:
[timer_test_stress_correction.cxx.txt](https://github.com/fltk/fltk/files/10950007/timer_test_stress_correction.cxx.txt)

**Comment by wcout** [comment]
@Albrecht-S I can confirm, that the patch given for `Fl_Timeout.cxx` makes it behave more like FLTK 1.3. But if this change was required to fix an other issue, I agree this will be no option.

Regarding your user code fix example: it is IMHO no solution, as  you just added an independent second timer, that has no effect, and changed the delay of the update timer to be the same as the simulated delay in the draw() method - so this does not have the "heavy draw load" symptom. If you change the sleep delay to more than 20ms the symptom is here again.

But I agree, that a solution should be done is user code (maybe using @MatthiasWM's proposal).

>  Therefore, event processing is kinda "preferred" over timeout processing, and in fact you can for instance wiggle the mouse over your window and thus stall timeout processing by continuously pushing events to the event queue. That would be hard to change. I have some ideas but I don't want to pursue this until it's really necessary.

I am not sure, if I like this. Does it mean a user can effect the timing by wiggling the mouse?

Maybe the different timeout handling of FLTK 1.4 could be described more detailed in the [migration](https://fltk.gitlab.io/fltk/migration_1_4.html) chapter ("Fl::timeout() and friends").

**Comment by Albrecht-S** [comment]
> Regarding your user code fix example: it is IMHO no solution, as you just added an independent second timer, that has no effect, and changed the delay of the update timer to be the same as the simulated delay in the draw() method - so this does not have the "heavy draw load" symptom. If you change the sleep delay to more than 20ms the symptom is here again.

@wcout Using the same delay as in the `draw()` method was by accident (you can change it with the `fps` constant), but you are probably right, this doesn't really resolve the problem of a long running `draw()` method. The only work around this is obviously to reduce the number of `draw()` calls which was my intention and worked in the given example as compared to the previous implementation of this specific demo program (in FLTK 1.4).

It may not solve your issue though, but there are other ways to do this as you noted above.

I wrote:
> Therefore, event processing is kinda "preferred" over timeout processing, ...

@wcout wrote:
> I am not sure, if I like this. Does it mean a user can effect the timing by 'wiggling the mouse'?

Yes, this is currently possible, but only if event processing in user code (for instance of `FL_MOVE` events) takes a **very** long time. This is usually not the case, but it is possible. I wrote a test program that shows this effect (using a delay of x milliseconds), and this looks indeed like a regression WRT 1.3.x. However, user code should NOT use much time in event handling, but I'm looking into this now, maybe I can improve this. But this should not keep this issue open.

Can we close this issue now?

**Comment by wcout** [comment]
> Can we close this issue now?

I am done with it as I know now the implications and a way to overcome the issue. Thanks for all clarifications and inputs.
If you think it is no longer needed (as a reminder) you may close it.

**Comment by Albrecht-S** [comment]
@wcout Thanks, I opened issue #698 to deal with further improvements of event and timeout processing, hence I'm closing this issue now.

FTR: I made a quick hack for Linux that "interrupts" event handling after a (variable) number of continuously processed events. This works around the "wiggling mouse" issue (which would stall timeout processing as described above) but has other negative effects (e.g. premature `Fl::flush(),` i.e. drawing although there are still pending events).
